feat: Add LoRA fine-tuning optimum-neuron example for slurm (#643)
aws-samples/awsome-distributed-training: 3.test_cases/pytorch/optimum-neuron/llama3/slurm/fine-tuning/README.md

> ## PEFT fine tuning of Llama 3 on SageMaker HyperPod ( trn1/trn1n)

Reviewer: ```suggestion
## PEFT fine tuning of Llama 3 on Slurm cluster ( trn1/trn1n)
```

### Prerequisites
This example showcases how to train llama 3 models using AWS Trainium instances and Huggingface Optimum Neuron. 🤗 Optimum Neuron is the interface between the 🤗 Transformers library and AWS Accelerators including AWS Trainium and AWS Inferentia. It provides a set of tools enabling easy model loading, training and inference on single- and multi-Accelerator settings for different downstream tasks.

> Before running this training, you'll need to create a SageMaker HyperPod cluster with at least 1 trn1.32xlarge/ trn1n.32xlarge instance group. Instructions can be found in the ﻿Cluster Setup﻿ section. 

Reviewer: ```suggestion
Before running this training, you'll need to create a slurm cluster with at least 1 trn1.32xlarge/ trn1n.32xlarge instance group. Instructions can be found in the ﻿Cluster Setup﻿ section. 
```

> Before running this training, you'll need to create a SageMaker HyperPod cluster with at least 1 trn1.32xlarge/ trn1n.32xlarge instance group. Instructions can be found in the ﻿Cluster Setup﻿ section. 

Reviewer: ```suggestion
Before running this training, you'll need to create a slurm cluster with at least 1 trn1.32xlarge/ trn1n.32xlarge instance group. Instructions can be found in the ﻿Cluster Setup﻿ section. 
```

> You will also need to complete the following prerequisites for configuring and deploying your SageMaker HyperPod cluster for fine tuning:
> - Submit a service quota increase request to get access to Trainium instances in your AWS Region. You will need to request an increase for Amazon EC2 Trn1 instances, ml.trn1.32xlarge or ml.trn1n.32xlarge.
> - Locally, install the AWS Command Line Interface (AWS CLI); the required minimum version needed is 2.14.3.
> - Locally, Install the AWS Systems Manager Session Manager Plugin in order to SSH into your cluster.

Reviewer: ```suggestion
```
Please refer https://github.com/aws-samples/awsome-distributed-training/tree/main/1.architectures or workshop instead.

Additionally, since Llama 3 is a gated model users have to register in Huangface and obtain an ﻿access token﻿ before running this example.

### Setup
In this section, we will setup our training environment on the cluster. Begin by logging into your cluster by following the ﻿SSH into Cluster﻿ section.

#### Step 2: Setup Python Environment

> Setup a virtual python environment and install your training dependencies. Make sure this repo is stored on the shared FSX volume of your cluster so all nodes have access to it.

Reviewer: ```suggestion
Setup a virtual python environment and install your training dependencies. Make sure this repo is stored on the shared volume of your cluster so all nodes have access to it.
```
It could be another file system (ex. OpenZFS).

```bash

> chmod +x submit_jobs/0_create_env.sh
> sbatch submit_jobs/0_create_env.sh

Reviewer: ```suggestion
bash submit_jobs/0_create_env.sh
```
No need to run this on a compute node, as this step dose not require Trainium.

Author: From our experiment on a completely newly created cluster, we do need to run this on the compute node, because these steps install the python3.9-dev. Otherwise we have been seeing this error when `import torch_xla`:
```
ImportError: libpython3.10.so.1.0: cannot open shared object file: No such file or directory
srun: error: ip-10-1-27-155: task 0: Exited with exit code 1
```
It was resolved by installing this library on the compute node.

Reviewer: I see. Then probably we want to ask to install python3.9 on all the trainium nodes. ex.
```bash
srun --partition=<partition_name> sudo apt-get update && sudo apt-get install -y python3.9
```
The current guide assumes that there is only one trn1.32xl instance deployed on a cluster, which is not always the case.

```
View the logs created by the scripts in this lab by running this command below. You can update it for the step you are currently running:

```
tail -f logs/0_create_env.out 
```
Before proceeding to the next step throughout this lab, check if the current job has finished by running:

```
squeue
```

### Training

#### Step 1: Compile the model
Before you begin training on Trainium with Neuron, you will need to pre-compile your model with the ﻿neuron_parallel_compile CLI﻿.  This will trace through the model's training code and apply optimizations to improve performance. 

```bash

> chmod +x 2_compile_model.sh
> sbatch 2_compile_model.sh

Reviewer: ```suggestion
sbatch bash 2_compile_model.sh
```

```
The compilation process will generate NEFF (Neuron Executable File Format) files that will speed up your model's fine tuning job.

#### Step 2: Fine Tuning
With your model compiled, you can now begin fine tuning your Llama 3 model. 

For the purposes of this workshop, we will use the ﻿dolly 15k dataset﻿. As part of the training process, the script below will download the dataset and format it into a way that the model expects. Each data point will contain an instruction that guides the model's task, optional context that provides background information, and response that represent the desired output.

Now submit the fine tuning job:

```bash

> chmod +x 3_finetune.sh
> sbatch 3_finetune.sh

Reviewer: ```suggestion
sbatch bash 3_finetune.sh
```

Author: Thank you, sounds good, all scripts are committed as executable.

```

#### Step 3: Model Weight Consolidation
After training has completed, you will have a new directory for your model checkpoints. This directory will contain the model checkpoint shards from each neuron device that were generated during training. Use the model consolidation script to combine the shards into a single `model.safetensor` file.

```bash

> chmod +x 4_model_consolidation.sh
> sbatch 4_model_consolidation.sh

Reviewer: ```suggestion
sbatch bash 4_model_consolidation.sh
```

```
The `model.safetensor` file will contain the LoRA weights of your model that were updated during training.

#### Step 4: Merge Lora Weights
After consolidating the model shards, merge the LoRA adapter weights back to your base Llama 3 model:

```bash

> chmod +x 5_merge_lora_weights.sh
> sbatch 5_merge_lora_weights.sh

Reviewer: ```suggestion
sbatch bash 5_merge_lora_weights.sh
```

```
Your final fine tuned model weights will be saved to the `final_model_path` directory.

#### Step 5: Validate your trained model
Now that your model is fine tuned, see how its generations differ from the base model for the dolly-15k dataset.

```bash

> chmod +x 6_inference.sh
> sbatch 6_inference.sh

Reviewer: ```suggestion
sbatch bash 6_inference.sh
```

You will also need to complete the following prerequisites for configuring and deploying your SageMaker HyperPod cluster for fine tuning:

* Submit a service quota increase request to get access to Trainium instances in your AWS Region. You will need to request an increase for Amazon EC2 Trn1 instances, ml.trn1.32xlarge or ml.trn1n.32xlarge.
* Locally, install the AWS Command Line Interface (AWS CLI); the required minimum version needed is 2.14.3.
* Locally, Install the AWS Systems Manager Session Manager Plugin in order to SSH into your cluster.

Additionally, since Llama 3 is a gated model users have to register in Huggingface and obtain an [access token](https://huggingface.co/docs/hub/en/security-tokens) before running this example.

### Training

## Step 1: Download training scripts

Begin by downloading the training scripts from the aws-awesome-distributed repo:

```bash
cd ~/
git clone https://github.com/aws-samples/awsome-distributed-training

mkdir ~/peft_ft 
cd ~/peft_ft

> cp -r ~/awsome-distributed-training/3.test_cases/pytorch/optimum-neuron/llama3/slurm/fine-tuning/submit_jobs .

Author: Yes, this is not necessary, I updated to just cd into this directory.
